File: scripts/fetch_chrome_only_apis.py
#!/usr/bin/env python3
"""Fetch and list WebExtension APIs supported in Chrome but not Firefox.

This script fetches the browser-compat-data from GitHub's API without cloning
the entire repository. It lists the webextensions/api directory and downloads
JSON files concurrently for fast analysis.

Usage:
    python3 scripts/fetch_chrome_only_apis.py

Requirements:
    pip install aiohttp  # For async HTTP requests
"""

import asyncio
import json
import sys
from pathlib import Path
from typing import Any, Dict, Iterator, Tuple, List, Union

try:
    import aiohttp  # type: ignore
except ImportError:
    print("Error: aiohttp is required for async operations", file=sys.stderr)
    print("Install with: pip install aiohttp", file=sys.stderr)
    sys.exit(1)

# GitHub repository info
REPO_OWNER = "mdn"
REPO_NAME = "browser-compat-data"
BRANCH = "main"
API_PATH = "webextensions/api"

# GitHub API URLs
GITHUB_API_BASE = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}"
GITHUB_RAW_BASE = f"https://raw.githubusercontent.com/{REPO_OWNER}/{REPO_NAME}/{BRANCH}"

PathParts = Tuple[str, ...]


async def fetch_json(session: aiohttp.ClientSession, url: str) -> Any:
    """Fetch JSON from URL asynchronously.
    
    Returns Any because different endpoints return different types:
    - GitHub contents API returns List[Dict]
    - Raw file content returns Dict
    """
    async with session.get(url) as response:
        response.raise_for_status()
        return await response.json(content_type=None)


async def list_api_files(session: aiohttp.ClientSession) -> List[str]:
    """List all JSON files in the webextensions/api directory."""
    url = f"{GITHUB_API_BASE}/contents/{API_PATH}?ref={BRANCH}"
    try:
        contents: List[Dict[str, Any]] = await fetch_json(session, url)
        return [
            item["name"]
            for item in contents
            if isinstance(item, dict) and item.get("name", "").endswith(".json")
        ]
    except Exception as e:
        print(f"Failed to list API files: {e}", file=sys.stderr)
        return []


async def fetch_api_file(session: aiohttp.ClientSession, filename: str) -> Tuple[str, Dict[str, Any]]:
    """Fetch a single API file from GitHub raw content."""
    url = f"{GITHUB_RAW_BASE}/{API_PATH}/{filename}"
    try:
        data = await fetch_json(session, url)
        return filename, data
    except Exception as e:
        print(f"Error fetching {filename}: {e}", file=sys.stderr)
        return filename, {}


def walk_support_entries(
    path: PathParts,
    node: Dict[str, Any],
    filename: str,
) -> Iterator[Tuple[PathParts, str, Dict[str, Any]]]:
    """Walk through compatibility entries in a data node."""
    compat = node.get("__compat")
    if isinstance(compat, dict):
        support = compat.get("support")
        if isinstance(support, dict):
            yield path, filename, support

    for key, child in node.items():
        if not isinstance(key, str) or key.startswith("__"):
            continue
        if isinstance(child, dict):
            yield from walk_support_entries(path + (key,), child, filename)
        elif isinstance(child, list):
            for index, item in enumerate(child):
                if isinstance(item, dict):
                    indexed_key = f"{key}[{index}]"
                    yield from walk_support_entries(path + (indexed_key,), item, filename)


def is_supported(entry: Any) -> bool:
    """Return True if the given support statement counts as supported."""
    if entry is None:
        return False
    if isinstance(entry, list):
        return any(is_supported(item) for item in entry)
    if isinstance(entry, bool):
        return entry
    if isinstance(entry, str):
        return entry.strip().lower() not in {"", "false", "no"}
    if isinstance(entry, dict):
        version_added = entry.get("version_added")
        if isinstance(version_added, str):
            return version_added.strip() not in {"", "false", "mirrored"}
        return bool(version_added)
    return False


def format_version(entry: Any) -> str:
    """Return a human-readable description of the browser support entry."""
    if entry is None:
        return "not supported"
    if isinstance(entry, list):
        formatted = [format_version(item) for item in entry]
        return "; ".join(formatted)
    if isinstance(entry, bool):
        return "supported" if entry else "not supported"
    if isinstance(entry, str):
        text = entry.strip() or "not supported"
        return text
    if isinstance(entry, dict):
        version_added = entry.get("version_added")
        if isinstance(version_added, str):
            version = version_added.strip()
            return version or "not supported"
        if version_added in (None, False):
            return "not supported"
        return str(version_added)
    return str(entry)


async def process_api_files(session: aiohttp.ClientSession, api_files: List[str]) -> List[Tuple[str, str, Any, Any]]:
    """Process all API files concurrently and return Chrome-only APIs."""
    results = []
    total = len(api_files)
    
    # Fetch all files concurrently
    print(f"Fetching {total} files concurrently...", file=sys.stderr)
    tasks = [fetch_api_file(session, filename) for filename in api_files]
    
    # Wait for all downloads to complete
    file_data = await asyncio.gather(*tasks, return_exceptions=True)
    
    # Process downloaded data
    print(f"Processing {total} files...", file=sys.stderr)
    processed = 0
    
    for item in file_data:
        # Handle exceptions from gather
        if isinstance(item, Exception):
            print(f"Skipping file due to error: {item}", file=sys.stderr)
            continue
        
        # Type guard: item is Tuple[str, Dict[str, Any]] at this point
        filename, data = item  # type: Tuple[str, Dict[str, Any]]
        processed += 1
        
        if processed % 10 == 0:
            print(f"Processed {processed}/{total} files...", file=sys.stderr)
        
        if not data:
            continue
            
        api_section = data.get("webextensions", {}).get("api", {})
        if not isinstance(api_section, dict):
            continue
        
        for api_name, api_data in api_section.items():
            if not isinstance(api_data, dict):
                continue
            
            for path_parts, source_file, support in walk_support_entries(
                (api_name,), api_data, filename
            ):
                chrome_info = support.get("chrome")
                firefox_info = support.get("firefox")
                
                if is_supported(chrome_info) and not is_supported(firefox_info):
                    feature_path = ".".join(path_parts)
                    results.append((feature_path, source_file, chrome_info, firefox_info))
    
    print(f"Completed processing all {processed} files", file=sys.stderr)
    return results


async def main_async() -> None:
    """Main async function."""
    async with aiohttp.ClientSession() as session:
        print("Fetching API file list from GitHub...", file=sys.stderr)
        api_files = await list_api_files(session)
        
        if not api_files:
            print("No API files found.", file=sys.stderr)
            return
        
        print(f"Found {len(api_files)} API files. Processing...", file=sys.stderr)
        results = await process_api_files(session, api_files)
        
        if not results:
            print("\nNo APIs found that are supported in Chrome but not in Firefox.")
            return
        
        print(f"\n{'='*80}")
        print("WebExtension APIs supported in Chrome but not Firefox:")
        print(f"{'='*80}\n")
        
        for api_name, source_file, chrome_info, firefox_info in sorted(results, key=lambda x: x[0].lower()):
            print(f"- {api_name}")
            print(f"    Source: {source_file}")
            print(f"    Chrome: {format_version(chrome_info)}")
            print(f"    Firefox: {format_version(firefox_info)}\n")
        
        print(f"\nTotal: {len(results)} Chrome-only APIs found")


def main() -> None:
    """Main entry point."""
    try:
        asyncio.run(main_async())
    except KeyboardInterrupt:
        print("\n\nInterrupted by user", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"\nError: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/scripts/fetch_chrome_only_apis.rs
use std::collections::HashSet;
use std::time::Duration;

use anyhow::{Context, Result};
use futures::{stream, StreamExt};
use reqwest::Client;
use serde::Deserialize;
use serde_json::{Map, Value};

use crate::parser::javascript::CHROME_ONLY_APIS;

const REPO_OWNER: &str = "mdn";
const REPO_NAME: &str = "browser-compat-data";
const BRANCH: &str = "main";
const API_PATH: &str = "webextensions/api";

const GITHUB_API_BASE: &str = "https://api.github.com/repos";
const GITHUB_RAW_BASE: &str = "https://raw.githubusercontent.com";

#[derive(Debug)]
struct ChromeOnlyApi {
    feature_path: String,
    source_file: String,
    chrome_info: Value,
    firefox_info: Value,
}

#[derive(Debug, Deserialize)]
struct ContentItem {
    name: String,
}

pub async fn run() -> Result<()> {
    let client = Client::builder()
        .user_agent("chrome-to-firefox (https://github.com/OtsoBear/chrome-to-firefox)")
        .timeout(Duration::from_secs(30))
        .build()
        .context("failed to build HTTP client")?;

    eprintln!("Fetching API file list from GitHub...");
    let api_files = list_api_files(&client).await?;

    if api_files.is_empty() {
        eprintln!("No API files found.");
        return Ok(());
    }

    eprintln!("Found {} API files. Processing...", api_files.len());
    let results = process_api_files(&client, &api_files).await?;

    if results.is_empty() {
        println!("\nNo APIs found that are supported in Chrome but not in Firefox.");
        return Ok(());
    }

    println!("\n{}", "=".repeat(80));
    println!("WebExtension APIs supported in Chrome but not Firefox:");
    println!("{}\n", "=".repeat(80));

    let mut sorted_results = results;
    sorted_results.sort_by(|a, b| a.feature_path.to_lowercase().cmp(&b.feature_path.to_lowercase()));

    let mut implemented_count = 0usize;
    let mut not_implemented_count = 0usize;
    let mut matched_prefixes: HashSet<&str> = HashSet::new();

    for entry in &sorted_results {
        let chrome_path = format!("chrome.{}", entry.feature_path);
        let status = if let Some(prefix) = matches_known_chrome_only(&chrome_path) {
            matched_prefixes.insert(prefix);
            implemented_count += 1;
            "[implemented]"
        } else {
            not_implemented_count += 1;
            "[not implemented]"
        };

        println!("- {} {}", entry.feature_path, status);
        println!("    Source: {}", entry.source_file);
        println!("    Chrome: {}", format_version(&entry.chrome_info));
        println!("    Firefox: {}\n", format_version(&entry.firefox_info));
    }

    println!("Summary:");
    println!("  Total Chrome-only APIs found: {}", sorted_results.len());
    println!("  Implemented (matches parser/javascript.rs): {}", implemented_count);
    println!("  Not yet implemented: {}", not_implemented_count);

    let missing_prefixes: Vec<&str> = CHROME_ONLY_APIS
        .iter()
        .copied()
        .filter(|prefix| !matched_prefixes.contains(prefix))
        .collect();

    println!(
        "  Known chrome-only prefixes tracked: {}",
        CHROME_ONLY_APIS.len()
    );
    println!(
        "  Known prefixes missing from MDN dataset: {}",
        missing_prefixes.len()
    );

    if !missing_prefixes.is_empty() {
        println!("    Missing prefixes:");
        for prefix in missing_prefixes {
            println!("      - {}", prefix);
        }
    }

    println!("");
    println!("Use this summary to prioritize new compatibility shims.");

    Ok(())
}

async fn list_api_files(client: &Client) -> Result<Vec<String>> {
    let url = format!(
        "{}/{}/{}/contents/{}?ref={}",
        GITHUB_API_BASE, REPO_OWNER, REPO_NAME, API_PATH, BRANCH
    );

    let response = client
        .get(url)
        .send()
        .await
        .context("failed to request API file list")?
        .error_for_status()
        .context("GitHub API returned an error for contents list")?;

    let items: Vec<ContentItem> = response
        .json()
        .await
        .context("failed to parse contents list")?;

    let files = items
        .into_iter()
        .filter(|item| item.name.ends_with(".json"))
        .map(|item| item.name)
        .collect();

    Ok(files)
}

async fn fetch_api_file(client: &Client, filename: &str) -> Result<Value> {
    let url = format!(
        "{}/{}/{}/{}/{}/{}",
        GITHUB_RAW_BASE, REPO_OWNER, REPO_NAME, BRANCH, API_PATH, filename
    );

    let response = client
        .get(url)
        .send()
        .await
        .with_context(|| format!("failed to download {filename}"))?
        .error_for_status()
        .with_context(|| format!("GitHub returned an error for {filename}"))?;

    response
        .json()
        .await
        .with_context(|| format!("failed to parse JSON for {filename}"))
}

async fn process_api_files(client: &Client, api_files: &[String]) -> Result<Vec<ChromeOnlyApi>> {
    let total = api_files.len();
    eprintln!("Fetching {} files concurrently...", total);

    let mut results = Vec::new();
    let mut processed = 0usize;

    let mut stream = stream::iter(api_files.iter().cloned())
        .map(|filename| {
            let client = client.clone();
            async move {
                let data = fetch_api_file(&client, &filename).await;
                (filename, data)
            }
        })
        .buffer_unordered(32);

    while let Some((filename, data)) = stream.next().await {
        match data {
            Ok(value) => {
                processed += 1;
                if processed % 10 == 0 || processed == total {
                    eprintln!("Processed {processed}/{total} files...");
                }
                collect_chrome_only_apis(&filename, &value, &mut results);
            }
            Err(err) => {
                eprintln!("Error fetching {filename}: {err:?}");
            }
        }
    }

    eprintln!("Completed processing all {processed} files");
    Ok(results)
}

fn collect_chrome_only_apis(filename: &str, data: &Value, results: &mut Vec<ChromeOnlyApi>) {
    let api_section = data
        .get("webextensions")
        .and_then(|v| v.get("api"))
        .and_then(Value::as_object);

    let Some(api_section) = api_section else {
        return;
    };

    for (api_name, api_data) in api_section {
        if !api_data.is_object() {
            continue;
        }

        let mut path = vec![api_name.clone()];
        walk_support_entries(&mut path, api_data, filename, results);
    }
}

fn walk_support_entries(
    path: &mut Vec<String>,
    node: &Value,
    filename: &str,
    results: &mut Vec<ChromeOnlyApi>,
) {
    if let Value::Object(map) = node {
        if let Some(Value::Object(compat)) = map.get("__compat") {
            if let Some(Value::Object(support)) = compat.get("support") {
                handle_support_entry(path, filename, support, results);
            }
        }

        for (key, child) in map {
            if key.starts_with("__") {
                continue;
            }

            match child {
                Value::Object(_) => {
                    path.push(key.clone());
                    walk_support_entries(path, child, filename, results);
                    path.pop();
                }
                Value::Array(items) => {
                    for (index, item) in items.iter().enumerate() {
                        if item.is_object() {
                            path.push(format!("{}[{}]", key, index));
                            walk_support_entries(path, item, filename, results);
                            path.pop();
                        }
                    }
                }
                _ => {}
            }
        }
    }
}

fn handle_support_entry(
    path: &[String],
    filename: &str,
    support: &Map<String, Value>,
    results: &mut Vec<ChromeOnlyApi>,
) {
    let chrome_info = support.get("chrome");
    let firefox_info = support.get("firefox");

    if let Some(chrome_info) = chrome_info {
        if is_supported(chrome_info) && !is_supported(firefox_info.unwrap_or(&Value::Null)) {
            let feature_path = path.join(".");
            results.push(ChromeOnlyApi {
                feature_path,
                source_file: filename.to_string(),
                chrome_info: chrome_info.clone(),
                firefox_info: firefox_info.cloned().unwrap_or(Value::Null),
            });
        }
    }
}

fn matches_known_chrome_only(chrome_path: &str) -> Option<&'static str> {
    CHROME_ONLY_APIS
        .iter()
        .copied()
        .find(|prefix| chrome_path.starts_with(prefix))
}

fn is_supported(entry: &Value) -> bool {
    match entry {
        Value::Null => false,
        Value::Bool(value) => *value,
        Value::Number(_) => true,
        Value::String(text) => {
            let normalized = text.trim().to_lowercase();
            !normalized.is_empty() && normalized != "false" && normalized != "no"
        }
        Value::Array(items) => items.iter().any(is_supported),
        Value::Object(map) => {
            match map.get("version_added") {
                Some(Value::Bool(value)) => *value,
                Some(Value::Number(_)) => true,
                Some(Value::String(text)) => {
                    let normalized = text.trim().to_lowercase();
                    !matches!(normalized.as_str(), "" | "false" | "mirrored")
                }
                Some(Value::Array(items)) => items.iter().any(is_supported),
                Some(Value::Object(obj)) => is_supported(&Value::Object(obj.clone())),
                Some(Value::Null) | None => false,
            }
        }
    }
}

fn format_version(entry: &Value) -> String {
    match entry {
        Value::Null => "not supported".into(),
        Value::Bool(true) => "supported".into(),
        Value::Bool(false) => "not supported".into(),
        Value::Number(num) => num.to_string(),
        Value::String(text) => {
            let trimmed = text.trim();
            if trimmed.is_empty() {
                "not supported".into()
            } else {
                trimmed.to_string()
            }
        }
        Value::Array(items) => {
            let parts: Vec<String> = items.iter().map(format_version).collect();
            parts.join("; ")
        }
        Value::Object(map) => {
            match map.get("version_added") {
                Some(Value::String(text)) => {
                    let trimmed = text.trim();
                    if trimmed.is_empty() {
                        "not supported".into()
                    } else {
                        trimmed.to_string()
                    }
                }
                Some(Value::Bool(true)) => "supported".into(),
                Some(Value::Bool(false)) | Some(Value::Null) | None => "not supported".into(),
                Some(Value::Number(num)) => num.to_string(),
                Some(other) => format_version(other),
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn supported_detection_handles_various_forms() {
        assert!(is_supported(&json!(true)));
        assert!(is_supported(&json!("1.0")));
        assert!(!is_supported(&json!("")));
        assert!(!is_supported(&json!("false")));
        assert!(is_supported(&json!("Note")));
        assert!(is_supported(&json!({"version_added": "50"})));
        assert!(!is_supported(&json!({"version_added": ""})));
        assert!(is_supported(&json!({"version_added": true})));
        assert!(!is_supported(&json!({"version_added": false})));
    }

    #[test]
    fn format_version_matches_expectations() {
        assert_eq!(format_version(&json!(null)), "not supported");
        assert_eq!(format_version(&json!(true)), "supported");
        assert_eq!(format_version(&json!("")), "not supported");
        assert_eq!(format_version(&json!(" 72 " )), "72");
        assert_eq!(
            format_version(&json!({"version_added": "42"})),
            "42"
        );
        assert_eq!(
            format_version(&json!([{ "version_added": "1" }, null])),
            "1; not supported"
        );
    }

    #[test]
    fn matches_known_prefixes() {
        assert!(matches_known_chrome_only("chrome.offscreen.createDocument").is_some());
        assert!(matches_known_chrome_only("chrome.action.openPopup").is_some());
        assert!(matches_known_chrome_only("chrome.tabs.getSelected").is_some());
        assert!(matches_known_chrome_only("chrome.runtime.getPackageDirectoryEntry").is_some());
    }

    #[test]
    fn detects_unknown_prefixes() {
        assert!(matches_known_chrome_only("chrome.tabs.query").is_none());
        assert!(matches_known_chrome_only("chrome.cookies.getAll").is_none());
    }
}
